aesd-char-driver/main.c

```diff
@@ -0,0 +1,286 @@
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/syscalls.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include <asm/uaccess.h>
+#include "aesdchar.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Jordi Cros Mompart");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    //Handle open
+    PDEBUG("open\n");
+    try_module_get(THIS_MODULE);
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    //Handle release
+    PDEBUG("release\n");
+    module_put(THIS_MODULE);
+
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t  entry_offset = 0;
+    struct aesd_buffer_entry *entry = NULL;
+    PDEBUG("read %zu bytes with offset %lld\n",count,*f_pos);
+    
+    //"count" is the number of bytes to be returned
+    //"f_pos" is the offset from where to start the read
+
+    //Set mutex
+    mutex_lock(&aesd_device.lock);
+
+    //Get the entry corresponding to the request
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&aesd_device.buffer, *f_pos, &entry_offset);
+    //Check for the results
+    if(!entry)
+    {
+        PDEBUG("Access to aesdchar is returning without contents read\n");
+        mutex_unlock(&aesd_device.lock);
+        *f_pos = 0;
+        retval = 0;
+        return retval;
+    }
+    //Copy from entry offset, "count" or size
+    if(entry_offset + count >= entry->size)
+    {
+        //Copy only until the size
+        if(copy_to_user(buf, &entry->buffptr[entry_offset],  entry->size - entry_offset) != 0)
+        {
+            PDEBUG("Could not copy memory to the user\n");
+            mutex_unlock(&aesd_device.lock);
+            retval = -EINVAL;
+            return retval;
+        }
+        *f_pos = *f_pos + entry->size;
+        retval = entry->size - entry_offset;
+    }
+    else
+    {
+        //Copy until the count
+        //Copy only until the size
+        if(copy_to_user(buf, &entry->buffptr[entry_offset],  count) != 0)
+        {
+            PDEBUG("Could not copy memory to the user\n");
+            mutex_unlock(&aesd_device.lock);
+            retval = -EINVAL;
+            return retval;
+        }
+        *f_pos = *f_pos + count;
+        retval = count;
+    }
+
+    //Release mutex
+    mutex_unlock(&aesd_device.lock);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    const char *to_be_freed = NULL;
+    PDEBUG("write %zu bytes with offset %lld\n",count,*f_pos);
+    
+
+    //Check filp and buf validity
+    if(!filp || !buf)
```

Comment:
Good defensive programming here but I think in this case you can assume the file operations setup in the kernel will not give you a null `filp`. See reference implementation in [scull](https://github.com/cu-ecen-aeld/ldd3/blob/master/scull/main.c#L339) for the assumptions which are typically made.